File: src/config/bot-config.ts
import * as dotenv from 'dotenv';
import * as os from 'os';

export interface BotConfigData {
    polymarket_api_key: string;
    polymarket_api_secret: string;
    polymarket_host: string;
    gamma_host: string;
    starting_capital: number;
    position_size: number;
    min_profit_threshold: number;
    min_market_volume: number;
    market_scan_limit: number;
    scan_interval_ms: number;
    execution_delay_ms: number;
    paper_trading: boolean;
    mock_mode: boolean;
    log_level: string;
    gamma_rps: number;
}

export class BotConfig implements BotConfigData {
    polymarket_api_key!: string;
    polymarket_api_secret!: string;
    polymarket_host!: string;
    gamma_host!: string;
    starting_capital!: number;
    position_size!: number;
    min_profit_threshold!: number;
    min_market_volume!: number;
    market_scan_limit!: number;
    scan_interval_ms!: number;
    execution_delay_ms!: number;
    paper_trading!: boolean;
    mock_mode!: boolean;
    log_level!: string;
    gamma_rps!: number;

    constructor(data: BotConfigData) {
        Object.assign(this, data);
    }

    static load(): BotConfig {
        dotenv.config();

        const getEnvBool = (name: string, defaultValue: boolean): boolean => {
            const val = process.env[name];
            if (val === undefined) return defaultValue;
            return ["1", "true", "yes", "y", "on"].includes(val.toLowerCase());
        };

        const getEnvFloat = (name: string, defaultValue: number): number => {
            const val = process.env[name];
            if (val === undefined || val === "") return defaultValue;
            return parseFloat(val);
        };

        const getEnvInt = (name: string, defaultValue: number): number => {
            const val = process.env[name];
            if (val === undefined || val === "") return defaultValue;
            return parseInt(val, 10);
        };

        return new BotConfig({
            polymarket_api_key: process.env.POLYMARKET_API_KEY || "",
            polymarket_api_secret: process.env.POLYMARKET_API_SECRET || "",
            polymarket_host: process.env.POLYMARKET_HOST || "https://clob.polymarket.com",
            gamma_host: process.env.GAMMA_HOST || "https://gamma-api.polymarket.com",
            starting_capital: getEnvFloat("STARTING_CAPITAL", getEnvFloat("CAPITAL", 100.0)),
            position_size: getEnvFloat("POSITION_SIZE", 10.0),
            min_profit_threshold: getEnvFloat("MIN_PROFIT_THRESHOLD", 0.005),
            min_market_volume: getEnvFloat("MIN_MARKET_VOLUME", 10000.0),
            market_scan_limit: getEnvInt("MARKET_SCAN_LIMIT", 1000),
            scan_interval_ms: getEnvInt("SCAN_INTERVAL", 30000),
            execution_delay_ms: getEnvInt("EXECUTION_DELAY", 1000),
            paper_trading: getEnvBool("PAPER_TRADING", true),
            mock_mode: getEnvBool("MOCK_MODE", true),
            log_level: process.env.LOG_LEVEL || "info",
            gamma_rps: getEnvInt("GAMMA_RPS", 5),
        });
    }

    validate(): void {
        if (this.starting_capital <= 0) throw new Error("STARTING_CAPITAL must be > 0");
        if (this.position_size <= 0) throw new Error("POSITION_SIZE must be > 0");
        if (this.min_profit_threshold <= 0) throw new Error("MIN_PROFIT_THRESHOLD must be > 0");
        if (this.scan_interval_ms <= 0) throw new Error("SCAN_INTERVAL must be > 0");
        if (this.execution_delay_ms < 0) throw new Error("EXECUTION_DELAY must be >= 0");
        if (this.market_scan_limit <= 0) throw new Error("MARKET_SCAN_LIMIT must be > 0");
    }
}

File: src/core/bot_config.py
from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Optional

from dotenv import load_dotenv


def _getenv_bool(name: str, default: bool) -> bool:
    raw = os.getenv(name)
    if raw is None:
        return default
    return raw.strip().lower() in {"1", "true", "yes", "y", "on"}


def _getenv_float(name: str, default: float) -> float:
    raw = os.getenv(name)
    if raw is None or raw == "":
        return default
    return float(raw)


def _getenv_int(name: str, default: int) -> int:
    raw = os.getenv(name)
    if raw is None or raw == "":
        return default
    return int(raw)


@dataclass(frozen=True)
class BotConfig:
    """Config for the pure arbitrage bot."""

    polymarket_api_key: str
    polymarket_api_secret: str
    polymarket_host: str
    gamma_host: str

    starting_capital: float
    position_size: float
    min_profit_threshold: float
    min_market_volume: float

    market_scan_limit: int
    scan_interval_ms: int
    execution_delay_ms: int

    paper_trading: bool
    mock_mode: bool
    log_level: str

    gamma_rps: int

    @classmethod
    def load(cls, dotenv_path: Optional[str] = None) -> "BotConfig":
        load_dotenv(dotenv_path=dotenv_path)

        return cls(
            polymarket_api_key=os.getenv("POLYMARKET_API_KEY", ""),
            polymarket_api_secret=os.getenv("POLYMARKET_API_SECRET", ""),
            polymarket_host=os.getenv("POLYMARKET_HOST", "https://clob.polymarket.com"),
            gamma_host=os.getenv("GAMMA_HOST", "https://gamma-api.polymarket.com"),
            starting_capital=_getenv_float("STARTING_CAPITAL", _getenv_float("CAPITAL", 100.0)),
            position_size=_getenv_float("POSITION_SIZE", 10.0),
            min_profit_threshold=_getenv_float("MIN_PROFIT_THRESHOLD", 0.005),
            min_market_volume=_getenv_float("MIN_MARKET_VOLUME", 10_000.0),
            market_scan_limit=_getenv_int("MARKET_SCAN_LIMIT", 1000),
            scan_interval_ms=_getenv_int("SCAN_INTERVAL", 30_000),
            execution_delay_ms=_getenv_int("EXECUTION_DELAY", 1_000),
            paper_trading=_getenv_bool("PAPER_TRADING", True),
            mock_mode=_getenv_bool("MOCK_MODE", True),
            log_level=os.getenv("LOG_LEVEL", "info"),
            gamma_rps=_getenv_int("GAMMA_RPS", 5),
        )

    def validate(self) -> None:
        if self.starting_capital <= 0:
            raise ValueError("STARTING_CAPITAL must be > 0")
        if self.position_size <= 0:
            raise ValueError("POSITION_SIZE must be > 0")
        if self.min_profit_threshold <= 0:
            raise ValueError("MIN_PROFIT_THRESHOLD must be > 0")
        if self.scan_interval_ms <= 0:
            raise ValueError("SCAN_INTERVAL must be > 0")
        if self.execution_delay_ms < 0:
            raise ValueError("EXECUTION_DELAY must be >= 0")
        if self.market_scan_limit <= 0:
            raise ValueError("MARKET_SCAN_LIMIT must be > 0")
